### ISSUE TITLE
memory.free_dirty_pages in Waterfox 56.x

### ISSUE THREAD

**Issue by grahamperrin** [BrowserWorks/waterfox#423]
Consider making `memory.free_dirty_pages` `true` by default. 

I have tested with and without this for a few weeks. 

Qualitatively, I can **not** confirm that things are better when it's `true` … however my use cases are unusual (more on this in a separate issue), so I should not draw any conclusion. 

## Favouring `true` by default

From Mozilla bug [1414150 - Remove some unnecessary prefs in and around XPCOM](https://bugzilla.mozilla.org/show_bug.cgi?id=1414150) (RESOLVED FIXED in Firefox 58): 

>>> … "memory.free_dirty_pages" pref
>>>
>>> There's no good reason why someone would want this disabled.

>> Please, is the same true for 56.x? 
>> 
>> I'm interested from a Waterfox 56.x perspective. 

> I believe it is true. Disabling that pref will increase memory usage, while being extremely unlikely to provide any other benefits.

https://www.reddit.com/r/firefox/search?q=memory.free_dirty_pages&restrict_sr=on finds one match but truly there are at least two mentions: 

- https://www.reddit.com/r/firefox/comments/642j8g/unlike_the_last_version_of_firefox_when_available/dg1goy3/ under _Unlike the last version of firefox, when available RAM is exhausted, closing tabs doesn't free up RAM. Are there settings that can be changed to address this?_
- https://www.reddit.com/r/firefox/comments/7inrim/firefox_57_constantly_swapping_after_ram_is_full/dr09aoz/ under _Firefox 57 constantly swapping after RAM is full_

https://www.reddit.com/r/waterfox/search?q=memory.free_dirty_pages&restrict_sr=on finds nothing but truly there's at least one mention: 

- https://www.reddit.com/r/waterfox/comments/7m9zk7/how_have_you_optimized_waterfox/drthv4e/ under _How have you optimized waterfox?_

[betterwebleon/international-list: Cosmetic and tracking filters for uBlock Origin and uBlock.](https://github.com/betterwebleon/international-list)

[Linux Tricks: How to speed up the Firefox?](http://www.linuxtricks.info/2014/04/how-to-speed-up-firefox.html#post-body-7692733223383656414) (2014-04)

## Thought-provoking

<https://superuser.com/questions/789548/why-does-my-firefox-memory-usage-keep-rising-with-use-and-never-returns-to-the-i#comment1305525_789551> under _Why does my Firefox memory usage keep rising with use and never returns to the initial level? - Super User_: 

> You seem to be arguing that it's okay for Firefox to bloat up to 2Gb because that memory is just swap and so other programs are not prevented from having that 2Gb. This is false. Most of that 2Gb is dirty pages which have to be flushed. Out. Moreover, when Firefox decides to run some garbage collection on that cruft, it has to be paged back in: and that will happen in some random order that causes seeks all over the place on a conventional spinning hard drive. Not writing 2Gb to disk and reading it back is demonstrably more performant than doing so.
> 
> _show 31 more comments_

## Pages with no mention of _dirty_

[Reducing memory usage - Firefox - MozillaZine Knowledge Base](http://kb.mozillazine.org/Reducing_memory_usage_%28Firefox%29) (2013-05-11)

[List of all the Config provided by Configuration Mania](https://cat-in-136.github.io/2017/10/list-of-all-the-config-provided-by-configuration-mania.html) (2017-10-26)

## Other possible points of reference

[Bug 805855 - Free dirty pages in response to all memory-pressure messages. r=jlebar (fbc8afdb) · Commits · mer-core / gecko-dev · GitLab](https://git.merproject.org/mer-core/gecko-dev/commit/fbc8afdb64aee983b3fb8d5980ea2e1fa4f0d8c2)

https://git.merproject.org/search?search=memory.free_dirty_pages&group_id=&project_id=709&search_code=true&repository_ref=master

[ToDo: diffs FF57-FF58 · Issue #302 · ghacksuserjs/ghacks-user.js](https://github.com/ghacksuserjs/ghacks-user.js/issues/302)

**Comment by MrAlex94** [comment]
Thanks for investigating this Graham! 

So, I feel like I'm going a bit crazy here. Looking at Waterfox's git commit history, [this](https://github.com/MrAlex94/Waterfox/commit/2513a08f856bb476e83dec7bb3f4ea78c2155813) commit seems to have already been part, yet I've just had to commit it [manually](https://github.com/MrAlex94/Waterfox/commit/dd9203049651b7e5b9a478a4bc9ac24d4af2f5d0) again. How odd!

Thanks once again :-)

**Comment by grahamperrin** [comment]
I toggled the preference on numerous occasions, at times when I _wished_ for a positive difference. Typically not long after closing a tab in which I used Riot. 

To be honest, I never noticed a difference, so I just got on with things without adding noise to this issue. My use cases are pretty weird so I can't draw conclusions :-) 

FWIW I'll stick with _false_, and a mental note to experiment when things progress beyond 56.2.x

Cheers!

**Comment by grahamperrin** [comment]
```text
$ date ; uname -v
Wed 16 May 2018 11:28:11 BST
FreeBSD 12.0-CURRENT #2 r333587: Mon May 14 01:59:02 BST 2018     root@momh167-gjp4-hpelitebook8570p-freebsd:/usr/obj/usr/src/amd64.amd64/sys/GENERIC 
$ pkg info waterfox | grep Version
Version        : 56.2.0.7
```

First run of two profiles, with very few tabs, both felt somehow slower than usual, so I revisited this issue and (now looking at the commit) I see that the preference is removed and AFAICT will have no effect if added. 

I'll quit both instances then try a more methodical approach to get a feel of whether this build 20180515213058 is truly slower …

**Comment by grahamperrin** [comment]
> … First run of two profiles, with very few tabs, both felt somehow slower than usual, … get a feel of whether this build 20180515213058 is truly slower …

Happily: ater a couple of days, I can _not_ say that things are worse without the `memory.free_dirty_pages` preference. 

tl;dr a few minutes ago with a ~941 tab session I closed a three-tab window that included memory-hungry Riot. There followed a period of Waterfox not responding (to be expected in my environment), and: 

- it wasn't too long; and then 
- surprisingly, use of memory dropped. 

That drop, in good time, is quite unlike past experiences of closing a _tab_ to Riot. Things _do_ feel better, but that might be because recently I've not done much to force use of swap. 

Parallel to what's above: there may have been enhancements to Riot; not long ago I updated to a _much_ more recent version of FreeBSD-CURRENT, which might have better management of memory and/or better tuning of ZFS; and so on. I'll not over-analyse things. Hopefully I'll have nothing more to add to this issue. Drawing a line under it –

----
